File: frontend/src/types/digitalPresenceLink.ts
export type DigitalPlatform =
  | 'GITHUB'
  | 'LINKEDIN'
  | 'PERSONAL_PORTFOLIO'
  | 'GOOGLE_SCHOLAR'
  | 'RESEARCHGATE'
  | 'BEHANCE'
  | 'DRIBBBLE'
  | 'KAGGLE'
  | 'DEVPOST'
  | 'OTHER';

export type DigitalPresenceCategory =
  | 'TECHNICAL'
  | 'PROFESSIONAL'
  | 'ACADEMIC'
  | 'CREATIVE'
  | 'OTHER';

export const ADMISSION_VALUE_NOTE_MAX_LENGTH = 1000;

export const PLATFORM_LABELS: Record<DigitalPlatform, string> = {
  GITHUB: 'GitHub',
  LINKEDIN: 'LinkedIn',
  PERSONAL_PORTFOLIO: 'Personal Portfolio',
  GOOGLE_SCHOLAR: 'Google Scholar',
  RESEARCHGATE: 'ResearchGate',
  BEHANCE: 'Behance',
  DRIBBBLE: 'Dribbble',
  KAGGLE: 'Kaggle',
  DEVPOST: 'Devpost',
  OTHER: 'Other',
};

export const CATEGORY_LABELS: Record<DigitalPresenceCategory, string> = {
  TECHNICAL: 'Technical',
  PROFESSIONAL: 'Professional',
  ACADEMIC: 'Academic',
  CREATIVE: 'Creative',
  OTHER: 'Other',
};

export const PLATFORM_DEFAULT_CATEGORY: Record<DigitalPlatform, DigitalPresenceCategory> = {
  GITHUB: 'TECHNICAL',
  KAGGLE: 'TECHNICAL',
  DEVPOST: 'TECHNICAL',
  LINKEDIN: 'PROFESSIONAL',
  PERSONAL_PORTFOLIO: 'PROFESSIONAL',
  GOOGLE_SCHOLAR: 'ACADEMIC',
  RESEARCHGATE: 'ACADEMIC',
  BEHANCE: 'CREATIVE',
  DRIBBBLE: 'CREATIVE',
  OTHER: 'OTHER',
};

export const PLATFORM_OPTIONS: DigitalPlatform[] = [
  'GITHUB',
  'LINKEDIN',
  'PERSONAL_PORTFOLIO',
  'GOOGLE_SCHOLAR',
  'RESEARCHGATE',
  'BEHANCE',
  'DRIBBBLE',
  'KAGGLE',
  'DEVPOST',
  'OTHER',
];

export const CATEGORY_OPTIONS: DigitalPresenceCategory[] = [
  'TECHNICAL',
  'PROFESSIONAL',
  'ACADEMIC',
  'CREATIVE',
  'OTHER',
];

const URL_PATTERN =
  /^https?:\/\/(?:localhost(?::\d{2,5})?|(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,63})(?::\d{2,5})?(?:\/[^\s]*)?$/i;

export function normalizeWebUrl(value: string): string {
  const cleaned = value.trim();
  if (!cleaned) {
    throw new Error('URL is required.');
  }
  const withScheme = /^https?:\/\//i.test(cleaned) ? cleaned : `https://${cleaned}`;
  if (!URL_PATTERN.test(withScheme)) {
    throw new Error('Enter a valid web URL (include a domain such as example.com).');
  }
  return withScheme;
}

export interface DigitalPlatformOption {
  value: DigitalPlatform;
  label: string;
  default_category: DigitalPresenceCategory;
}

export interface DigitalPresenceCategoryOption {
  value: DigitalPresenceCategory;
  label: string;
}

export interface DigitalPresenceLinkRecord {
  id: number;
  platform_name: DigitalPlatform | null;
  platform_label: string | null;
  url: string | null;
  category: DigitalPresenceCategory | null;
  category_label: string | null;
  admission_value_note: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface DigitalPresenceLinksResponse {
  booking_id?: number | null;
  lead_id: number | null;
  platform_options: DigitalPlatformOption[];
  category_options: DigitalPresenceCategoryOption[];
  links: DigitalPresenceLinkRecord[];
  saved_at: string | null;
}

export interface DigitalPresenceLinkFormState {
  platform_name: DigitalPlatform | '';
  url: string;
  category: DigitalPresenceCategory | '';
  admission_value_note: string;
}

export const emptyDigitalPresenceLinkForm = (): DigitalPresenceLinkFormState => ({
  platform_name: '',
  url: '',
  category: '',
  admission_value_note: '',
});

export function linkToForm(record: DigitalPresenceLinkRecord): DigitalPresenceLinkFormState {
  return {
    platform_name: record.platform_name ?? '',
    url: record.url ?? '',
    category: record.category ?? '',
    admission_value_note: record.admission_value_note ?? '',
  };
}

export function formToLinkPayload(form: DigitalPresenceLinkFormState) {
  return {
    platform_name: form.platform_name || null,
    url: form.url.trim() ? normalizeWebUrl(form.url) : null,
    category: form.category || null,
    admission_value_note: form.admission_value_note.trim() || null,
  };
}

export function validateDigitalPresenceLinkForm(
  form: DigitalPresenceLinkFormState
): Record<string, string> {
  const errors: Record<string, string> = {};

  if (!form.platform_name) {
    errors.platform_name = 'Platform is required.';
  }
  if (!form.url.trim()) {
    errors.url = 'URL is required.';
  } else {
    try {
      normalizeWebUrl(form.url);
    } catch (err) {
      errors.url = err instanceof Error ? err.message : 'Enter a valid web URL.';
    }
  }
  if (!form.category) {
    errors.category = 'Category is required.';
  }
  if (form.admission_value_note.length > ADMISSION_VALUE_NOTE_MAX_LENGTH) {
    errors.admission_value_note = `Value to admission must be ${ADMISSION_VALUE_NOTE_MAX_LENGTH} characters or fewer.`;
  }

  return errors;
}

export const FALLBACK_PLATFORM_OPTIONS: DigitalPlatformOption[] = PLATFORM_OPTIONS.map(value => ({
  value,
  label: PLATFORM_LABELS[value],
  default_category: PLATFORM_DEFAULT_CATEGORY[value],
}));

export const FALLBACK_CATEGORY_OPTIONS: DigitalPresenceCategoryOption[] = CATEGORY_OPTIONS.map(
  value => ({
    value,
    label: CATEGORY_LABELS[value],
  })
);

File: backend/app/schemas/digital_presence_link.py
from __future__ import annotations

import re
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field, field_validator


class DigitalPlatform(str, Enum):
    GITHUB = "GITHUB"
    LINKEDIN = "LINKEDIN"
    PERSONAL_PORTFOLIO = "PERSONAL_PORTFOLIO"
    GOOGLE_SCHOLAR = "GOOGLE_SCHOLAR"
    RESEARCHGATE = "RESEARCHGATE"
    BEHANCE = "BEHANCE"
    DRIBBBLE = "DRIBBBLE"
    KAGGLE = "KAGGLE"
    DEVPOST = "DEVPOST"
    OTHER = "OTHER"


class DigitalPresenceCategory(str, Enum):
    TECHNICAL = "TECHNICAL"
    PROFESSIONAL = "PROFESSIONAL"
    ACADEMIC = "ACADEMIC"
    CREATIVE = "CREATIVE"
    OTHER = "OTHER"


PLATFORM_LABELS: dict[DigitalPlatform, str] = {
    DigitalPlatform.GITHUB: "GitHub",
    DigitalPlatform.LINKEDIN: "LinkedIn",
    DigitalPlatform.PERSONAL_PORTFOLIO: "Personal Portfolio",
    DigitalPlatform.GOOGLE_SCHOLAR: "Google Scholar",
    DigitalPlatform.RESEARCHGATE: "ResearchGate",
    DigitalPlatform.BEHANCE: "Behance",
    DigitalPlatform.DRIBBBLE: "Dribbble",
    DigitalPlatform.KAGGLE: "Kaggle",
    DigitalPlatform.DEVPOST: "Devpost",
    DigitalPlatform.OTHER: "Other",
}

CATEGORY_LABELS: dict[DigitalPresenceCategory, str] = {
    DigitalPresenceCategory.TECHNICAL: "Technical",
    DigitalPresenceCategory.PROFESSIONAL: "Professional",
    DigitalPresenceCategory.ACADEMIC: "Academic",
    DigitalPresenceCategory.CREATIVE: "Creative",
    DigitalPresenceCategory.OTHER: "Other",
}

PLATFORM_DEFAULT_CATEGORY: dict[DigitalPlatform, DigitalPresenceCategory] = {
    DigitalPlatform.GITHUB: DigitalPresenceCategory.TECHNICAL,
    DigitalPlatform.KAGGLE: DigitalPresenceCategory.TECHNICAL,
    DigitalPlatform.DEVPOST: DigitalPresenceCategory.TECHNICAL,
    DigitalPlatform.LINKEDIN: DigitalPresenceCategory.PROFESSIONAL,
    DigitalPlatform.PERSONAL_PORTFOLIO: DigitalPresenceCategory.PROFESSIONAL,
    DigitalPlatform.GOOGLE_SCHOLAR: DigitalPresenceCategory.ACADEMIC,
    DigitalPlatform.RESEARCHGATE: DigitalPresenceCategory.ACADEMIC,
    DigitalPlatform.BEHANCE: DigitalPresenceCategory.CREATIVE,
    DigitalPlatform.DRIBBBLE: DigitalPresenceCategory.CREATIVE,
    DigitalPlatform.OTHER: DigitalPresenceCategory.OTHER,
}

CATEGORY_SORT_ORDER: dict[DigitalPresenceCategory, int] = {
    DigitalPresenceCategory.TECHNICAL: 1,
    DigitalPresenceCategory.PROFESSIONAL: 2,
    DigitalPresenceCategory.ACADEMIC: 3,
    DigitalPresenceCategory.CREATIVE: 4,
    DigitalPresenceCategory.OTHER: 5,
}

PLATFORM_SORT_ORDER: dict[DigitalPlatform, int] = {
    DigitalPlatform.GITHUB: 1,
    DigitalPlatform.LINKEDIN: 2,
    DigitalPlatform.PERSONAL_PORTFOLIO: 3,
    DigitalPlatform.GOOGLE_SCHOLAR: 4,
    DigitalPlatform.RESEARCHGATE: 5,
    DigitalPlatform.BEHANCE: 6,
    DigitalPlatform.DRIBBBLE: 7,
    DigitalPlatform.KAGGLE: 8,
    DigitalPlatform.DEVPOST: 9,
    DigitalPlatform.OTHER: 10,
}

_URL_PATTERN = re.compile(
    r"^https?://"
    r"(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+[A-Z]{2,63}|localhost)"
    r"(?::\d{2,5})?"
    r"(?:/[^\s]*)?$",
    re.IGNORECASE,
)


def normalize_web_url(value: str) -> str:
    cleaned = (value or "").strip()
    if not cleaned:
        raise ValueError("URL is required.")
    if not cleaned.startswith(("http://", "https://")):
        cleaned = f"https://{cleaned}"
    if not _URL_PATTERN.match(cleaned):
        raise ValueError("Enter a valid web URL (include a domain such as example.com).")
    return cleaned


class DigitalPlatformOption(BaseModel):
    value: DigitalPlatform
    label: str
    default_category: DigitalPresenceCategory


class DigitalPresenceCategoryOption(BaseModel):
    value: DigitalPresenceCategory
    label: str


def list_platform_options() -> list[DigitalPlatformOption]:
    return [
        DigitalPlatformOption(
            value=platform,
            label=PLATFORM_LABELS[platform],
            default_category=PLATFORM_DEFAULT_CATEGORY[platform],
        )
        for platform in sorted(PLATFORM_SORT_ORDER, key=lambda item: PLATFORM_SORT_ORDER[item])
    ]


def list_category_options() -> list[DigitalPresenceCategoryOption]:
    return [
        DigitalPresenceCategoryOption(value=category, label=CATEGORY_LABELS[category])
        for category in sorted(CATEGORY_SORT_ORDER, key=lambda item: CATEGORY_SORT_ORDER[item])
    ]


class DigitalPresenceLinkInput(BaseModel):
    platform_name: DigitalPlatform | None = None
    url: str | None = Field(default=None, max_length=500)
    category: DigitalPresenceCategory | None = None
    admission_value_note: str | None = Field(default=None, max_length=1000)

    @field_validator("url")
    @classmethod
    def validate_url(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        if not stripped:
            return None
        return normalize_web_url(stripped)

    @field_validator("admission_value_note")
    @classmethod
    def normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class DigitalPresenceLinkOut(BaseModel):
    id: int
    platform_name: DigitalPlatform | None
    platform_label: str | None
    url: str | None
    category: DigitalPresenceCategory | None
    category_label: str | None
    admission_value_note: str | None
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DigitalPresenceLinksResponse(BaseModel):
    booking_id: int | None = None
    lead_id: int | None
    platform_options: list[DigitalPlatformOption]
    category_options: list[DigitalPresenceCategoryOption]
    links: list[DigitalPresenceLinkOut]
    saved_at: datetime | None = None
